ID persistence

pajot opened the pull request:
We copy tags from shape.text to shape.name at template instantiation to make them resistant to user mischief :)

lysnikolaou committed:
Fix typo in .travis.yml

lysnikolaou committed:
Update .travis.yml

pajot committed:
Merge branch 'master' of https://github.com/timhoffm/pptx-blueprint

pajot committed:
Merge branch 'master' of https://github.com/timhoffm/pptx-blueprint

pajot committed:
Merge remote-tracking branch 'upstream/master'

pajot committed:
Add _get_all_shapes internal method

Add _get_all_shapes internal method. The test simply checks to see if
the attributes we expect to use the shape for are present and whether
they can be coerced to strings.

pajot committed:
Copy tags to shape.name attribute for persistence

Copy tags at template instantiation from shape.text to shape.name
to make them resistant to user clobbering.

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -82,6 +84,20 @@ def _find_shapes_in_slide(slide):
+    def _copy_tags_to_name(self) -> None:
+        all_shapes = self._get_all_shapes()
+        # This regex matches on tags
+        regex_tag = re.compile(r'\{[\s\w]*\}')
```
We should be more strict on the the user-supplied tags.

1. The tag id must match  \w+ (no spaces).
2. We can be permissive and allow whitespace around the tag. - Allowing these can gracefully handle leading or trailing spaces, which might have slipped in unnoticed when editing the template.
3. We shouldn't suuport tags with other text, e.g. `some {tag}` or worse `{tag1} {tab2}`. - At least for now. Handling such things reasonably is much more difficult.

```suggestion
        regex_tag = re.compile(r'^\s*(\{\w+\})\s*$')
```

pajot replied:
That makes sense. I've applied your suggestions, I hope the PR is usable now :)

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -82,6 +84,20 @@ def _find_shapes_in_slide(slide):
+        for shape in all_shapes:
+            # We only copy contents we recognize as tags
+            if regex_tag.match(shape.text):
+                shape.name = shape.text
```
```suggestion
                shape.name = regex_tag.group(1)
```
We should write exactly the tag into the `name`.

Not quite sure if this is better with or without curly braces. Pro-with: It's explicit and there is little chance of mixing this up with existing names of non-placeholder shapes. Con-with: It's more verbose and more cumbersome to match in `_find_shapes`.

pajot replied:
I would leave the curly braces. How else can we distinguish it from names not meant to be tags?

pajot committed:
Update pptx_blueprint/__init__.py

Co-Authored-By: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

pajot committed:
Update pptx_blueprint/__init__.py

Co-Authored-By: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -82,6 +84,20 @@ def _find_shapes_in_slide(slide):
+        regex_tag = re.compile(r'^\s*(\{\w+\})\s*$')
+        for shape in all_shapes:
+            # We only copy contents we recognize as tags
+            if regex_tag.match(shape.text):
```
The group is stored in the match object:
~~~
match = regex_tag.match(shape.text)
if match:
    shape.name = match.group(1)
~~~

pajot committed:
Merge branch 'master' of https://github.com/timhoffm/pptx-blueprint into id_persistence

pajot committed:
Merge branch 'id_persistence' of https://github.com/pajot/pptx-blueprint into id_persistence

pajot committed:
Use the regex match group for shape.name

Use the regex match group for shape.name. This way we also include the curly
braces and ignore non-matching text in shape.text.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -87,6 +89,21 @@ def _find_shapes_in_slide(slide):
         return matched_shapes
 
+    def _get_all_shapes(self) -> Iterable[BaseShape]:
+        # Do we need all the shapes? Perhaps we should filter on tags here.
```
We need to iterate over each shape at least once and apply the re.match to identify whether there's some work for us to do. It's a decision of style if you want to first get all shapes and then match and work, or if you get all matching slides and then do some work. For a (extremely) big presentation, however, generating a list of all slides may be quite memory expensive and possibly slow. Both issues could be eliminated if you'd just replace the square brackets with regular brackets, thereby changing the list comprehension into a generator expression.

cmahr replied:
Also, please inline...

pajot replied:
> Also, please inline...

Do you mean the comment? Isn't it a bit long to be inlined? (PEP8 says "use inline comments sparingly ;) )

timhoffm replied:
Inlining means to eliminate the function and use the code of the function body instead of the function call. This is reasonable in this particular case because

- the function is only used once (and unlikely to be needed multiple times)
- the inline version is almost equally short and equally descriptive.
  instead of
  ~~~
  all_shapes = self._get_all_shapes()
  ~~~
  just use
  ~~~
  all_shapes = (shape
                for slide in self._presentation.slides
                for shape in slide.shapes)
  ~~~

cmahr replied:
Thanks for clarifying my comment, Tim. I now see that I could have been a little more verbose :-)

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -87,6 +89,21 @@ def _find_shapes_in_slide(slide):
+        return all_shapes
+
+    def _copy_tags_to_name(self) -> None:
+        all_shapes = self._get_all_shapes()
```
please inline

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -87,6 +89,21 @@ def _find_shapes_in_slide(slide):
+    def _copy_tags_to_name(self) -> None:
+        all_shapes = self._get_all_shapes()
+        # This regex matches on tags
+        regex_tag = re.compile(r'^\s*(\{\w+\})\s*$')
```
Could this be defined as a class constant?

cmahr replied:
Further, why are the curly brackets part of the group? Do we really want to set the shape.name to e.g. '{location}' instead of 'location'?

pajot replied:
> Further, why are the curly brackets part of the group? Do we really want to set the shape.name to e.g. '{location}' instead of 'location'?

Tim and I had discussed this, yes. Leaving the curly braces in makes it clear that this is a tag.

timhoffm replied:
> Could this be defined as a class constant?

 Could do, but if so, please private. I don't want to make this user-configurable as of now. Strictly speaking, it would even be sufficient not to explicitly compile the regexp and just use inline `if re.match(r'^\s*(\{\w+\})\s*$', shape.text):`. `re` caches the last few (I belive to remember 100?) expressions. Therefore, the overhead for this basic call vs. the compiled regexp is just a dict lookup, which is negligible.

> Do we really want to set the shape.name to e.g. '{location}' instead of 'location'?

I tend to say yes. While the curly brace feels a bit bulky here, it serves to highly reduce the likelihood of a name collision between a tag and an existing shape name.

cmahr commented on tests/test_template.py:
```diff
@@ -38,3 +39,16 @@ def test_find_shapes_from_one_slide(template):
+def test_get_all_shapes(template):
+    shapes = template._get_all_shapes()
+    for shape in shapes:
+        assert str(shape.text) and str(shape.name)
```
So how does this test that I 'get all the shapes'? I'm afraid that you'd have to mock the `_presentation` to verify this...

pajot replied:
I simply named the test after the function it is testing. But no, it is not actually testing that I have all the shapes. But code to test this is necessarily going to be as complex as the code that performs it.

At the very least, the code checks that what was obtained has at least some properties we would expect to find in slides :)

cmahr replied:
Yes, the test would be rather complex. At least you'd have to monkey patch the `Presentation` class (because we don't use dependency injection) and configure the mock (which seems tedious). Further, there are people out there that say "don't test private methods, just test the public interface", i.e., you would not write a test for `_get_all_shapes()` at all, but only test it indirectly by testing the public method it is used in.

Would probably the simplest to just give the test a more descriptive (cool slang for "longer") name ;-)

cmahr commented on tests/test_template.py:
```diff
@@ -38,3 +39,16 @@ def test_find_shapes_from_one_slide(template):
+    regex_tag = re.compile(r'\{[\s\w]*\}')
+    for shape in all_shapes:
+        if regex_tag.match(shape.text):
+            assert shape.text == shape.name
```
Is this really what you wan't to test? Taking a look at the regex you defined in the implementation, for a shape having the text '  {location}  ', you would set the name as '{location}'. Hence, in this case, `shape.text != shape.name`.

pajot replied:
You're right, this test needs to be fixed :)

pajot commented on pptx_blueprint/__init__.py:
```diff
@@ -71,7 +73,7 @@ def _find_shapes(self,
 
         def _find_shapes_in_slide(slide):
-            return filter(lambda shape: shape.text == f'{{{tag_name}}}', slide.shapes)
+            return filter(lambda shape: shape.name == f'{{{tag_name}}}', slide.shapes)
```
Sorry, I don't understand this change. I thought we wanted users to be able to place tags on the slide. shape.name is an object property.
